src/main/java/com/aws/greengrass/mqtt/bridge/clients/LocalMqtt5Client.java

```diff
@@ -0,0 +1,353 @@
+                    .withSubscription(topic, QOS).build();
+            LOGGER.atDebug().kv(TOPIC, topic).log("Subscribing to MQTT topic");
+            //TODO: deal with subscribeFuture
+            this.subscribeFuture = client.subscribe(subscribePacket);
+        }
+    }
+
+    @SuppressWarnings("PMD.AvoidCatchingGenericException")
+    private void subscribeToTopics(Set<String> topics) {
+        // TODO: Support configurable qos
+        // retry until interrupted
+        topics.forEach(s -> {
+            try {
+                RetryUtils.runWithRetry(mqttExceptionRetryConfig, () -> {
+                    subscribe(s);
+                    // useless return
+                    return null;
+                }, "subscribe-mqtt5-topic", LOGGER);
+                subscribedLocalMqttTopics.add(s);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            } catch (Exception e) {
+                LOGGER.atError().setCause(e).kv(TOPIC, s).log("Failed to subscribe");
+            }
+        });
+    }
+
+    private void resubscribe() {
+        subscribedLocalMqttTopics.clear();
+        // Resubscribe to topics
+        updateSubscriptionsInternal();
+    }
+
+    private void unsubscribe(String topic) {
+        if (client.getIsConnected()) {
+            UnsubscribePacket unsubscribePacket =
+                    new UnsubscribePacket.UnsubscribePacketBuilder().withSubscription(topic).build();
+            LOGGER.atDebug().kv(TOPIC, topic).log("Unsubscribing from MQTT topic");
+            //TODO: deal with unsubscribeFuture
+            CompletableFuture<UnsubAckPacket> unsubscribeFuture = client.unsubscribe(unsubscribePacket);
+        }
+    }
+
+    private synchronized void removeMappingAndSubscriptions() {
+        unsubscribeAll();
+        subscribedLocalMqttTopics.clear();
+    }
+
+    private void unsubscribeAll() {
+        LOGGER.atDebug().kv("mapping", subscribedLocalMqttTopics).log("Unsubscribe from local MQTT topics");
+
+        this.subscribedLocalMqttTopics.forEach(s -> {
+            unsubscribe(s);
+        });
+    }
+
+    @Override
+    public void start() {
+        try {
+            connectAndSubscribe();
+        } catch (CrtRuntimeException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private synchronized void connectAndSubscribe() throws CrtRuntimeException {
+        LOGGER.atInfo()
+                .kv(BridgeConfig.KEY_BROKER_URI, brokerUri)
+                .kv(BridgeConfig.KEY_CLIENT_ID, clientId)
+                .log("Connecting to broker");
+        reconnectAndResubscribe();
+    }
+
+    private synchronized void doConnect() throws CrtRuntimeException {
+        if (!client.getIsConnected()) {
+            connectFuture = new CompletableFuture<>();
+            client.start();
+        }
+    }
+
+    private void reconnectAndResubscribe() {
+        doConnect();
+        resubscribe();
```

Comment:
so looks like you have a connect future, we'll need to wait on it before resubscribing.  instead, could also subscribe within the connect callback, if we can avoid having an extra future variable to maintain that'd be good